uefi/src/proto/network/http.rs

```diff
@@ -0,0 +1,384 @@
+            ..Default::default()
+        };
+
+        let config = HttpConfigData {
+            http_version: HttpVersion::HTTP_VERSION_10,
+            time_out_millisec: 10_000,
+            local_addr_is_ipv6: false.into(),
+            access_point: HttpAccessPoint { ipv4_node: &ip4 },
+        };
+
+        self.protocol.as_mut().unwrap().configure(&config)?;
+        debug!("http: configure ok");
+
+        Ok(())
+    }
+
+    /// Send HTTP request
+    pub fn request(
+        &mut self,
+        method: HttpMethod,
+        url: &str,
+        body: Option<&mut [u8]>,
+    ) -> uefi::Result<()> {
+        let url16 = uefi::CString16::try_from(url).unwrap();
+
+        let Some(hostname) = url.split('/').nth(2) else {
+            return Err(Status::INVALID_PARAMETER.into());
+        };
+        let mut c_hostname = String::from(hostname);
+        c_hostname.push('\0');
+        debug!("http: host: {}", hostname);
+
+        let mut tx_req = HttpRequestData {
+            method,
+            url: url16.as_ptr().cast::<u16>(),
+        };
+
+        let mut tx_hdr = Vec::new();
+        tx_hdr.push(HttpHeader {
+            field_name: c"Host".as_ptr().cast::<u8>(),
+            field_value: c_hostname.as_ptr(),
+        });
+
+        let mut tx_msg = HttpMessage::default();
+        tx_msg.data.request = &mut tx_req;
+        tx_msg.header_count = tx_hdr.len();
+        tx_msg.header = tx_hdr.as_mut_ptr();
+        if body.is_some() {
+            let b = body.unwrap();
+            tx_msg.body_length = b.len();
+            tx_msg.body = b.as_mut_ptr().cast::<c_void>();
+        }
+
+        let mut tx_token = HttpToken {
+            status: Status::NOT_READY,
+            message: &mut tx_msg,
+            ..Default::default()
+        };
+
+        let p = self.protocol.as_mut().unwrap();
+        p.request(&mut tx_token)?;
+        debug!("http: request sent ok");
+
+        loop {
+            if tx_token.status != Status::NOT_READY {
+                break;
+            }
+            p.poll()?;
+        }
+
+        if tx_token.status != Status::SUCCESS {
+            return Err(tx_token.status.into());
+        };
+
+        debug!("http: request status ok");
+
+        Ok(())
+    }
+
+    /// Send HTTP GET request
+    pub fn request_get(&mut self, url: &str) -> uefi::Result<()> {
+        self.request(HttpMethod::GET, url, None)?;
+        Ok(())
+    }
+
+    /// Send HTTP HEAD request
+    pub fn request_head(&mut self, url: &str) -> uefi::Result<()> {
+        self.request(HttpMethod::HEAD, url, None)?;
+        Ok(())
+    }
+
+    /// Receive the start of the http response, the headers and (parts of) the body.
+    pub fn response_first(&mut self, expect_body: bool) -> uefi::Result<HttpHelperResponse> {
+        let mut rx_rsp = HttpResponseData {
+            status_code: HttpStatusCode::STATUS_UNSUPPORTED,
+        };
+
+        let mut body = vec![0; if expect_body { 16 * 1024 } else { 0 }];
```

Comment:
16 * 1024? Where do these numbers come from?

Reply:
16k buffer for the body.  Figured by experimenting a bit.  I havn't seen the uefi protocol return more in one call.  Often it is much less though, between 1000 + 1500 (payload of one ethernet frame?).